packages/remix-dev/compiler/compileBrowser.ts

```diff
@@ -123,29 +152,108 @@ export const createBrowserCompiler = (
   options: CompileOptions
 ): BrowserCompiler => {
-  let compiler: esbuild.BuildIncremental;
-  let esbuildConfig = createEsbuildConfig(remixConfig, options);
+  let appCompiler: esbuild.BuildIncremental;
+  let cssCompiler: esbuild.BuildIncremental;
+
   let compile = async (manifestChannel: WriteChannel<AssetsManifest>) => {
-    let metafile: esbuild.Metafile;
-    if (compiler === undefined) {
-      compiler = await esbuild.build({
-        ...esbuildConfig,
-        metafile: true,
-        incremental: true,
-      });
-      invariant(compiler.metafile, "Expected metafile to be defined");
-      metafile = compiler.metafile;
-    } else {
-      let rebuild = await compiler.rebuild();
-      invariant(rebuild.metafile, "Expected metafile to be defined");
-      metafile = rebuild.metafile;
-    }
-    let manifest = await createAssetsManifest(remixConfig, metafile);
+    let appBuildTask = async () => {
+      appCompiler = await (!appCompiler
+        ? esbuild.build({
+            ...createEsbuildConfig("app", remixConfig, options),
+            metafile: true,
+            incremental: true,
+          })
+        : appCompiler.rebuild());
+
+      invariant(
+        appCompiler.metafile,
+        "Expected app compiler metafile to be defined"
+      );
+    };
+
+    let cssBuildTask = async () => {
+      if (!remixConfig.future.unstable_cssModules) {
+        return;
+      }
+
+      // The types aren't great when combining write: false and incremental: true
+      //  so we need to assert that it's an incremental build
+      cssCompiler = (await (!cssCompiler
+        ? esbuild.build({
+            ...createEsbuildConfig("css", remixConfig, options),
+            metafile: true,
+            incremental: true,
+            write: false,
+          })
+        : cssCompiler.rebuild())) as esbuild.BuildIncremental;
+
+      invariant(
+        cssCompiler.metafile,
+        "Expected CSS compiler metafile to be defined"
+      );
+
+      let cssBundlePath: string | undefined;
+      let outputFiles = cssCompiler.outputFiles || [];
+
+      await Promise.all(
+        outputFiles.map(async (outputFile) => {
+          let outputPath = outputFile.path;
+
+          // Only write bundled CSS and source map to disk
+          if (
+            path.dirname(outputPath) === remixConfig.assetsBuildDirectory &&
+            path.basename(outputPath).startsWith("css-bundle")
+          ) {
+            if (outputPath.endsWith(".css")) {
+              // Grab the CSS bundle path so we can use it to generate the manifest
+              cssBundlePath = outputPath;
+              await fse.ensureDir(path.dirname(outputPath));
+
+              let contents =
+                options.mode === "production"
+                  ? await postcss([
+                      // We need to discard duplicate rules since "composes"
+                      // in CSS Modules can result in duplicate styles
+                      postcssDiscardDuplicates(),
```

Comment:
Since we're balking on sourcemaps for the moment, is there any reason not to dedupe in development as well?

Reply:
I was thinking the same thing. I'll look into this.

Reply:
I got this working. I had to rework the `compileBrowser` logic a bit so that we can read esbuild's CSS source map and then pass it to PostCSS. It actually cleaned up the code quite a bit since we're now writing the CSS bundle and its source map in a single step rather than mapping over all output files.

Reply:
To be clear, we're still skipping out on the broken CSS Modules source maps, but we're maintaining the CSS source map that we get from esbuild, otherwise every rule is said to come from `css-bundle-xxxxx.css` rather than the individual built CSS files. Still not as good as it could be, but much better than nothing.